testutil/garbage/garbage.go

```diff
@@ -0,0 +1,304 @@
+// of 1 for every scalar kind and 10 for the recursive kinds.
+func DefaultInitialWeights() map[datamodel.Kind]int {
+	return map[datamodel.Kind]int{
+		datamodel.Kind_List:   10,
+		datamodel.Kind_Map:    10,
+		datamodel.Kind_Bool:   1,
+		datamodel.Kind_Bytes:  1,
+		datamodel.Kind_Float:  1,
+		datamodel.Kind_Int:    1,
+		datamodel.Kind_Link:   1,
+		datamodel.Kind_Null:   1,
+		datamodel.Kind_String: 1,
+	}
+}
+
+// DefaultWeights provides the default map of weights that can be overridden by
+// the Weights option. The default is an equal weighting of 1 for every kind.
+func DefaultWeights() map[datamodel.Kind]int {
+	return map[datamodel.Kind]int{
+		datamodel.Kind_List:   1,
+		datamodel.Kind_Map:    1,
+		datamodel.Kind_Bool:   1,
+		datamodel.Kind_Bytes:  1,
+		datamodel.Kind_Float:  1,
+		datamodel.Kind_Int:    1,
+		datamodel.Kind_Link:   1,
+		datamodel.Kind_Null:   1,
+		datamodel.Kind_String: 1,
+	}
+}
+
+// InitialWeights sets a per-kind weighting for the root node. That is, the weights
+// set here will determine the liklihood of the returned Node's direct .Kind().
+// These weights are ignored after the top-level Node (for recursive kinds,
+// obviously for scalar kinds there is only a top-level Node).
+//
+// The default initial weights bias toward Map and List kinds, by a ratio of
+// 10:1—i.e. the recursive kinds are more likely to appear at the top-level.
+func InitialWeights(initialWeights map[datamodel.Kind]int) Option {
+	return func(o *Options) {
+		o.initialWeights = initialWeights
+	}
+}
+
+// Weights sets a per-kind weighting for nodes appearing throughout the returned
+// graph. When assembling a graph, these weights determine the liklihood that
+// a given kind will be selected for that node.
+//
+// A weight of 0 will turn that kind off entirely. So, for example, if you
+// wanted output data with no maps or bytes, then set both of those weights to
+// zero, leaving the rest >0 and do the same for InitialWeights.
+//
+// The default weights are set to 1—i.e. there is an equal liklihood that any of
+// the valid kinds will be selected for any point in the graph.
+//
+// This option is overridden by InitialWeights (which also has a default even
+// if not set explicitly) for the top-level node.
+func Weights(weights map[datamodel.Kind]int) Option {
+	return func(o *Options) {
+		o.weights = weights
+	}
+}
+
+// TargetBlockSize sets a very rough bias in number of bytes that the resulting
+// Node may consume when encoded (i.e. the block size). This is a very
+// approximate measure, but over enough repeated Garbage() calls, the resulting
+// Nodes, once encoded, should have a median that is somewhere in this vicinity.
+//
+// The default target block size is 100. This should be tuned in accordance with
+// the anticipated average block size of the system under test.
+func TargetBlockSize(blockSize uint64) Option {
+	return func(o *Options) {
+		o.blockSize = blockSize
+	}
+}
+
+func init() {
```

Comment:
OK, to be honest, I didn't realise you could do a static `var foo = thing` (probably because I'm used to `var` for uninitialised/default values and `:=` for assigned), I tried a bunch of things but figured `const` was the best I could do to initialise things..

I've fixed these now but get an import cycle if I try `generators`:

```
# github.com/ipld/go-ipld-prime/testutil/garbage [github.com/ipld/go-ipld-prime/testutil/garbage.test]
./garbage.go:35:5: initialization loop:
        /home/rvagg/go/src/github.com/ipld/go-ipld-prime/testutil/garbage/garbage.go:35:5: generators refers to
        /home/rvagg/go/src/github.com/ipld/go-ipld-prime/testutil/garbage/garbage.go:93:6: listGenerator refers to
        /home/rvagg/go/src/github.com/ipld/go-ipld-prime/testutil/garbage/garbage.go:59:6: generate refers to
        /home/rvagg/go/src/github.com/ipld/go-ipld-prime/testutil/garbage/garbage.go:35:5: generators
FAIL    github.com/ipld/go-ipld-prime/testutil/garbage [build failed]
```

which is fair enough; got any suggestions for doing this outside of an `init()`?

Reply:
Ah, I had not noticed this. Seems fine to leave it in the init func, then - perhaps comment that there's an init cycle due to the generate func.

Reply:
added comment